### PR TITLE
Removed formatter dependency from JsonResult and JsonViewComponentResult

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #2715 
@pranavkm @rynowak

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs [hunk 1]
```diff
@@ -81,80 +59,44 @@ public JsonResult(object value, IOutputFormatter formatter)
...
-                    // If the available user-configured formatters can't write this type, then fall back to the
-                    // 'global' one.
-                    formatter = formatterContext
-                        .HttpContext
-                        .RequestServices
-                        .GetRequiredService<JsonOutputFormatter>();
-
-                    // Run SelectFormatter again to try to choose a content type that this formatter can do.
-                    objectResult.SelectFormatter(formatterContext, new[] { formatter });
+                    jsonWriter.CloseOutput = false;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: Does this even matter?

File: src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs [hunk 2]
```diff
@@ -18,13 +15,11 @@ namespace Microsoft.AspNet.Mvc
     /// </summary>
     public class JsonResult : ActionResult
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: So much simpler

File: src/Microsoft.AspNet.Mvc.Extensions/ViewComponents/JsonViewComponentResult.cs [hunk 3]
```diff
@@ -74,11 +65,5 @@ public Task ExecuteAsync([NotNull] ViewComponentContext context)
             Execute(context);
             return Task.FromResult(true);
         }
-
-        private static JsonOutputFormatter ResolveFormatter(ViewComponentContext context)
-        {
-            var services = context.ViewContext.HttpContext.RequestServices;
-            return services.GetRequiredService<JsonOutputFormatter>();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Extensions/ViewComponents/JsonViewComponentResult.cs]: You should be able to remove this from the service registration now right?

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs [hunk 4]
```diff
@@ -81,80 +59,44 @@ public JsonResult(object value, IOutputFormatter formatter)
...
             }
             else
             {
-                objectResult.ContentTypes = ContentTypes;
+                if (contentTypeHeader.Encoding == null)
+                {
+                    // 1. Do not modify the user supplied content type
+                    // 2. Parse here to handle parameters apart from charset
+                    contentTypeHeader = MediaTypeHeaderValue.Parse(contentTypeHeader.ToString());
+                    contentTypeHeader.Encoding = Encoding.UTF8;
```
Comment (IDisposable) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: What if the content type header contains the encoding?

Comment (darrelmiller) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: Technically it shouldn't.  Despite its widespread usage, the charset parameter is not defined for application/json.  https://tools.ietf.org/html/rfc7159#section-11

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: @IDisposable, I don't understand your concern here. This line won't be hit if the content type header already contains the encoding.

File: src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs [hunk 5]
```diff
@@ -18,13 +15,11 @@ namespace Microsoft.AspNet.Mvc
     /// </summary>
     public class JsonResult : ActionResult
     {
-        /// <summary>
-        /// The list of content-types used for formatting when <see cref="ContentTypes"/> is null or empty.
-        /// </summary>
-        public static readonly IReadOnlyList<MediaTypeHeaderValue> DefaultContentTypes = new MediaTypeHeaderValue[]
+        private JsonSerializerSettings _serializerSettings;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: `readonly`?

File: src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs [hunk 6]
```diff
@@ -18,13 +15,11 @@ namespace Microsoft.AspNet.Mvc
...
+
+        private readonly MediaTypeHeaderValue DefaultContentType = new MediaTypeHeaderValue("application/json")
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: `static readonly`

File: test/Microsoft.AspNet.Mvc.Extensions.Test/ViewComponents/JsonViewComponentResultTest.cs [hunk 7]
```diff
@@ -18,29 +18,12 @@ namespace Microsoft.AspNet.Mvc
...
-            // Assert
-            Assert.Equal(expectedFormatter, result.Formatter);
-            Assert.Equal("1", new StreamReader(buffer).ReadToEnd());
-        }
-
-        [Fact]
         public void Execute_UsesFormatter_WithSpecifiedSerializerSettings()
         {
             // Arrange
+            var abcdIndentedUTF8Bytes
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Extensions.Test/ViewComponents/JsonViewComponentResultTest.cs]: Could you use strings instead? `Assert.Equal(abcdIndented, Encoding.UTF8.GetString(body.ToArray())` - easier to understand what's happening

**Comment by pranavkm** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Extensions/ViewComponents/JsonViewComponentResult.cs (hunk 3 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Extensions/ViewComponents/JsonViewComponentResult.cs]: No. JsonHelper fails without this.

File: src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs (hunk 1 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: spoke to @kichalla, it is needed so we don't close the underlying response stream. MVC 5 has the same behavior.

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs (hunk 4 above)
Comment (IDisposable) [inline src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs]: Sorry, didn't read up higher. Reading @darrelmiller link it looks like it should be assumed to be UTF-8 (and the only other valid encodings would be UTF16 or UTF32, which should have been specified) https://tools.ietf.org/html/rfc7159#section-8.1

**Pull request merged**